File: skill-bridge/backend/src/main/java/com/skillbridge/service/contract/ProjectCloseRequestService.java
package com.skillbridge.service.contract;

import com.skillbridge.dto.contract.request.*;
import com.skillbridge.dto.contract.response.ProjectCloseRequestDetailDTO;
import com.skillbridge.dto.contract.response.ProjectCloseRequestResponse;
import com.skillbridge.entity.auth.User;
import com.skillbridge.entity.contract.*;
import com.skillbridge.repository.auth.UserRepository;
import com.skillbridge.repository.contract.*;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.Optional;

/**
 * Project Close Request Service
 * Handles business logic for project close requests
 * Story-41: Project Close Request for SOW Contract
 */
@Service
@Transactional
public class ProjectCloseRequestService {
    
    @Autowired
    private ProjectCloseRequestRepository projectCloseRequestRepository;
    
    @Autowired
    private SOWContractRepository sowContractRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private ContractHistoryRepository contractHistoryRepository;
    
    /**
     * Validate if a Close Request can be created for a SOW
     */
    public void validateCanCreateCloseRequest(Integer sowId, User currentUser) {
        // Check user role
        if (!"SALES_REP".equals(currentUser.getRole()) && !"SALES_MANAGER".equals(currentUser.getRole())) {
            throw new RuntimeException("Only Sales Representatives and Sales Managers can create close requests.");
        }
        
        // Check SOW exists
        SOWContract sow = sowContractRepository.findById(sowId)
            .orElseThrow(() -> new RuntimeException("SOW Contract not found"));
        
        // Check SOW status is Active
        if (sow.getStatus() != SOWContract.SOWContractStatus.Active) {
            throw new RuntimeException("Close request can only be created for Active SOW contracts.");
        }
        
        // Check SOW end_date <= today
        if (sow.getPeriodEnd() != null && sow.getPeriodEnd().isAfter(LocalDate.now())) {
            throw new RuntimeException("Close request can only be created when the contract end date has passed.");
        }
        
        // Check no existing Pending Close Request exists
        boolean hasPending = projectCloseRequestRepository.existsBySowIdAndStatus(
            sowId, 
            ProjectCloseRequest.ProjectCloseRequestStatus.Pending
        );
        if (hasPending) {
            throw new RuntimeException("A pending close request already exists for this SOW. Please wait for client response or resubmit the existing request.");
        }
    }
    
    /**
     * Create a new Project Close Request
     */
    public ProjectCloseRequestResponse createCloseRequest(
        Integer sowId, 
        CreateProjectCloseRequestRequest request, 
        User currentUser
    ) {
        // Validate
        validateCanCreateCloseRequest(sowId, currentUser);
        
        // Validate message and links length
        if (request.getMessage() != null && request.getMessage().length() > 5000) {
            throw new RuntimeException("Message must not exceed 5000 characters.");
        }
        if (request.getLinks() != null && request.getLinks().length() > 2000) {
            throw new RuntimeException("Links must not exceed 2000 characters.");
        }
        
        // Create entity
        ProjectCloseRequest closeRequest = new ProjectCloseRequest();
        closeRequest.setSowId(sowId);
        closeRequest.setRequestedByUserId(currentUser.getId());
        closeRequest.setStatus(ProjectCloseRequest.ProjectCloseRequestStatus.Pending);
        closeRequest.setMessage(request.getMessage());
        closeRequest.setLinks(request.getLinks());
        
        // Save
        closeRequest = projectCloseRequestRepository.save(closeRequest);
        
        // Create audit log
        createHistoryEntry(sowId, "CloseRequestCreated", 
            "Project close request created by " + currentUser.getFullName() + ".", 
            null, null, currentUser.getId());
        
        // Convert to DTO
        ProjectCloseRequestDetailDTO dto = convertToDetailDTO(closeRequest);
        
        return new ProjectCloseRequestResponse(true, "Project close request created successfully", dto);
    }
    
    /**
     * Get latest Close Request for a SOW
     */
    public ProjectCloseRequestDetailDTO getLatestCloseRequest(Integer sowId, User currentUser) {
        // Check SOW exists
        SOWContract sow = sowContractRepository.findById(sowId)
            .orElseThrow(() -> new RuntimeException("SOW Contract not found"));
        
        // Check access permission
        if ("CLIENT".equals(currentUser.getRole())) {
            // Client can only see their own SOWs
            if (!sow.getClientId().equals(currentUser.getId())) {
                throw new RuntimeException("Access denied: You do not have permission to view this close request.");
            }
        } else if ("SALES_REP".equals(currentUser.getRole()) || "SALES_MANAGER".equals(currentUser.getRole())) {
            // Sales Rep can only see assigned SOWs, Sales Manager can see all
            if ("SALES_REP".equals(currentUser.getRole())) {
                if (sow.getAssigneeUserId() == null || !sow.getAssigneeUserId().equals(currentUser.getId())) {
                    throw new RuntimeException("Access denied: You can only view close requests for contracts assigned to you");
                }
            }
        } else {
            throw new RuntimeException("Access denied: Invalid user role.");
        }
        
        // Get latest close request
        Optional<ProjectCloseRequest> closeRequestOpt = projectCloseRequestRepository
            .findFirstBySowIdOrderByCreatedAtDesc(sowId);
        
        if (closeRequestOpt.isEmpty()) {
            // Return null DTO instead of throwing exception - this is a valid state
            // The frontend will handle the case when no close request exists
            return null;
        }
        
        return convertToDetailDTO(closeRequestOpt.get());
    }
    
    /**
     * Approve a Close Request (Client action)
     */
    public ProjectCloseRequestResponse approveCloseRequest(Integer closeRequestId, User currentUser) {
        // Validate user role
        if (!"CLIENT".equals(currentUser.getRole())) {
            throw new RuntimeException("Only clients can approve close requests.");
        }
        
        // Get close request
        ProjectCloseRequest closeRequest = projectCloseRequestRepository.findById(closeRequestId)
            .orElseThrow(() -> new RuntimeException("Close request not found"));
        
        // Validate status
        if (closeRequest.getStatus() != ProjectCloseRequest.ProjectCloseRequestStatus.Pending) {
            throw new RuntimeException("This close request cannot be approved. Current status: " + closeRequest.getStatus());
        }
        
        // Get SOW
        SOWContract sow = sowContractRepository.findById(closeRequest.getSowId())
            .orElseThrow(() -> new RuntimeException("SOW Contract not found"));
        
        // Validate ownership
        if (!sow.getClientId().equals(currentUser.getId())) {
            throw new RuntimeException("You do not have permission to review this close request.");
        }
        
        // Update close request status
        closeRequest.setStatus(ProjectCloseRequest.ProjectCloseRequestStatus.ClientApproved);
        closeRequest = projectCloseRequestRepository.save(closeRequest);
        
        // Update SOW status to Completed
        sow.setStatus(SOWContract.SOWContractStatus.Completed);
        sow = sowContractRepository.save(sow);
        
        // Create audit log
        createHistoryEntry(sow.getId(), "CloseRequestApprovedByClient", 
            "Project close request approved by client. SOW marked as completed.", 
            null, null, currentUser.getId());
        
        // Convert to DTO
        ProjectCloseRequestDetailDTO dto = convertToDetailDTO(closeRequest);
        dto.setSowStatus("Completed");
        
        return new ProjectCloseRequestResponse(true, 
            "Project close request approved. SOW has been marked as completed.", dto);
    }
    
    /**
     * Reject a Close Request (Client action)
     */
    public ProjectCloseRequestResponse rejectCloseRequest(
        Integer closeRequestId, 
        RejectProjectCloseRequestRequest request, 
        User currentUser
    ) {
        // Validate user role
        if (!"CLIENT".equals(currentUser.getRole())) {
            throw new RuntimeException("Only clients can reject close requests.");
        }
        
        // Validate reason
        if (request.getReason() == null || request.getReason().trim().isEmpty()) {
            throw new RuntimeException("Rejection reason is required.");
        }
        if (request.getReason().length() > 2000) {
            throw new RuntimeException("Rejection reason must not exceed 2000 characters.");
        }
        
        // Get close request
        ProjectCloseRequest closeRequest = projectCloseRequestRepository.findById(closeRequestId)
            .orElseThrow(() -> new RuntimeException("Close request not found"));
        
        // Validate status
        if (closeRequest.getStatus() != ProjectCloseRequest.ProjectCloseRequestStatus.Pending) {
            throw new RuntimeException("This close request cannot be rejected. Current status: " + closeRequest.getStatus());
        }
        
        // Get SOW
        SOWContract sow = sowContractRepository.findById(closeRequest.getSowId())
            .orElseThrow(() -> new RuntimeException("SOW Contract not found"));
        
        // Validate ownership
        if (!sow.getClientId().equals(currentUser.getId())) {
            throw new RuntimeException("You do not have permission to review this close request.");
        }
        
        // Update close request
        closeRequest.setStatus(ProjectCloseRequest.ProjectCloseRequestStatus.Rejected);
        closeRequest.setClientRejectReason(request.getReason());
        closeRequest = projectCloseRequestRepository.save(closeRequest);
        
        // SOW status remains Active (no change)
        
        // Create audit log
        createHistoryEntry(sow.getId(), "CloseRequestRejectedByClient", 
            "Project close request rejected by client. Reason: " + request.getReason(), 
            null, null, currentUser.getId());
        
        // Convert to DTO
        ProjectCloseRequestDetailDTO dto = convertToDetailDTO(closeRequest);
        dto.setSowStatus("Active");
        
        return new ProjectCloseRequestResponse(true, 
            "Project close request rejected. SOW remains active.", dto);
    }
    
    /**
     * Resubmit a rejected Close Request (SalesRep action)
     */
    public ProjectCloseRequestResponse resubmitCloseRequest(
        Integer closeRequestId, 
        ResubmitProjectCloseRequestRequest request, 
        User currentUser
    ) {
        // Validate user role
        if (!"SALES_REP".equals(currentUser.getRole()) && !"SALES_MANAGER".equals(currentUser.getRole())) {
            throw new RuntimeException("Only Sales Representatives and Sales Managers can resubmit close requests.");
        }
        
        // Get close request
        ProjectCloseRequest closeRequest = projectCloseRequestRepository.findById(closeRequestId)
            .orElseThrow(() -> new RuntimeException("Close request not found"));
        
        // Validate status
        if (closeRequest.getStatus() != ProjectCloseRequest.ProjectCloseRequestStatus.Rejected) {
            throw new RuntimeException("Only rejected close requests can be resubmitted.");
        }
        
        // Validate ownership (original requester or SalesManager)
        if (!"SALES_MANAGER".equals(currentUser.getRole()) && 
            !closeRequest.getRequestedByUserId().equals(currentUser.getId())) {
            throw new RuntimeException("You do not have permission to resubmit this close request.");
        }
        
        // Validate message and links length
        if (request.getMessage() != null && request.getMessage().length() > 5000) {
            throw new RuntimeException("Message must not exceed 5000 characters.");
        }
        if (request.getLinks() != null && request.getLinks().length() > 2000) {
            throw new RuntimeException("Links must not exceed 2000 characters.");
        }
        
        // Update close request
        if (request.getMessage() != null) {
            closeRequest.setMessage(request.getMessage());
        }
        if (request.getLinks() != null) {
            closeRequest.setLinks(request.getLinks());
        }
        closeRequest.setStatus(ProjectCloseRequest.ProjectCloseRequestStatus.Pending);
        // Keep client_reject_reason for history (optional)
        closeRequest = projectCloseRequestRepository.save(closeRequest);
        
        // Create audit log
        createHistoryEntry(closeRequest.getSowId(), "CloseRequestResubmitted", 
            "Project close request resubmitted after rejection by " + currentUser.getFullName() + ".", 
            null, null, currentUser.getId());
        
        // Convert to DTO
        ProjectCloseRequestDetailDTO dto = convertToDetailDTO(closeRequest);
        
        return new ProjectCloseRequestResponse(true, 
            "Project close request resubmitted successfully", dto);
    }
    
    /**
     * Convert ProjectCloseRequest entity to DetailDTO
     */
    private ProjectCloseRequestDetailDTO convertToDetailDTO(ProjectCloseRequest closeRequest) {
        ProjectCloseRequestDetailDTO dto = new ProjectCloseRequestDetailDTO();
        dto.setId(closeRequest.getId());
        dto.setSowId(closeRequest.getSowId());
        dto.setStatus(closeRequest.getStatus().name());
        dto.setMessage(closeRequest.getMessage());
        dto.setLinks(closeRequest.getLinks());
        dto.setClientRejectReason(closeRequest.getClientRejectReason());
        dto.setCreatedAt(closeRequest.getCreatedAt());
        dto.setUpdatedAt(closeRequest.getUpdatedAt());
        
        // Load SOW information
        SOWContract sow = sowContractRepository.findById(closeRequest.getSowId())
            .orElse(null);
        if (sow != null) {
            dto.setSowContractName(sow.getContractName());
            dto.setSowStatus(sow.getStatus().name());
            // Format period
            if (sow.getPeriodStart() != null && sow.getPeriodEnd() != null) {
                dto.setSowPeriod(String.format("%s-%s", 
                    sow.getPeriodStart().toString().replace("-", "/"),
                    sow.getPeriodEnd().toString().replace("-", "/")));
            }
        }
        
        // Load requester information
        User requester = userRepository.findById(closeRequest.getRequestedByUserId())
            .orElse(null);
        if (requester != null) {
            ProjectCloseRequestDetailDTO.UserInfoDTO userInfo = 
                new ProjectCloseRequestDetailDTO.UserInfoDTO(
                    requester.getId(),
                    requester.getFullName(),
                    requester.getEmail()
                );
            dto.setRequestedBy(userInfo);
        }
        
        return dto;
    }
    
    /**
     * Create history entry for SOW contract
     */
    private void createHistoryEntry(Integer sowId, String activityType, String description, 
                                   String fileLink, String fileUrl, Integer createdBy) {
        ContractHistory history = new ContractHistory();
        history.setContractId(null); // SOW contracts use sow_contract_id
        history.setSowContractId(sowId);
        history.setHistoryType("SOW");
        history.setEntryDate(LocalDate.now());
        history.setDescription(description);
        history.setDocumentLink(fileLink);
        history.setDocumentName(fileUrl);
        history.setCreatedBy(createdBy);
        contractHistoryRepository.save(history);
    }
}

File: SEP490-SkillBridge-FE/src/services/projectCloseRequestService.ts
import { API_BASE_URL } from '@/lib/apiConfig';

/**
 * Project Close Request Service
 * Story-41: Project Close Request for SOW Contract
 */

export interface CreateProjectCloseRequestRequest {
    message?: string;
    links?: string;
}

export interface ResubmitProjectCloseRequestRequest {
    message?: string;
    links?: string;
}

export interface ApproveProjectCloseRequestRequest {
    confirm?: boolean;
}

export interface RejectProjectCloseRequestRequest {
    reason: string;
}

export interface UserInfo {
    id: number;
    name: string;
    email: string;
}

export interface ProjectCloseRequestDetail {
    id: number;
    sowId: number;
    sowContractName?: string;
    sowPeriod?: string;
    status: 'Pending' | 'ClientApproved' | 'Rejected';
    message?: string;
    links?: string;
    requestedBy?: UserInfo;
    clientRejectReason?: string;
    createdAt: string;
    updatedAt: string;
    sowStatus?: string;
}

export interface ProjectCloseRequestResponse {
    success: boolean;
    message: string;
    data: ProjectCloseRequestDetail;
}

/**
 * Create a new Project Close Request (SalesRep)
 * POST /api/sales/sows/{sowId}/close-requests
 */
export async function createProjectCloseRequest(
    sowId: number,
    request: CreateProjectCloseRequestRequest,
    token: string
): Promise<ProjectCloseRequestResponse> {
    const response = await fetch(`${API_BASE_URL}/sales/sows/${sowId}/close-requests`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Failed to create close request' }));
        throw new Error(error.message || 'Failed to create close request');
    }

    return response.json();
}

/**
 * Resubmit a rejected Project Close Request (SalesRep)
 * POST /api/sales/sows/{sowId}/close-requests/{id}/resubmit
 */
export async function resubmitProjectCloseRequest(
    sowId: number,
    closeRequestId: number,
    request: ResubmitProjectCloseRequestRequest,
    token: string
): Promise<ProjectCloseRequestResponse> {
    const response = await fetch(`${API_BASE_URL}/sales/sows/${sowId}/close-requests/${closeRequestId}/resubmit`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Failed to resubmit close request' }));
        throw new Error(error.message || 'Failed to resubmit close request');
    }

    return response.json();
}

/**
 * Get latest Project Close Request for a SOW (Client or SalesRep)
 * GET /api/client/sows/{sowId}/close-requests/latest
 * Returns null if no close request exists (404 response)
 */
export async function getLatestProjectCloseRequest(
    sowId: number,
    token: string
): Promise<ProjectCloseRequestDetail | null> {
    const response = await fetch(`${API_BASE_URL}/client/sows/${sowId}/close-requests/latest`, {
        method: 'GET',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
    });

    if (response.status === 404) {
        // No close request found - this is a valid state
        return null;
    }

    if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Failed to get close request' }));
        throw new Error(error.message || 'Failed to get close request');
    }

    return response.json();
}

/**
 * Approve a Project Close Request (Client)
 * POST /api/client/close-requests/{id}/approve
 */
export async function approveProjectCloseRequest(
    closeRequestId: number,
    request: ApproveProjectCloseRequestRequest,
    token: string
): Promise<ProjectCloseRequestResponse> {
    const response = await fetch(`${API_BASE_URL}/client/close-requests/${closeRequestId}/approve`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Failed to approve close request' }));
        throw new Error(error.message || 'Failed to approve close request');
    }

    return response.json();
}

/**
 * Reject a Project Close Request (Client)
 * POST /api/client/close-requests/{id}/reject
 */
export async function rejectProjectCloseRequest(
    closeRequestId: number,
    request: RejectProjectCloseRequestRequest,
    token: string
): Promise<ProjectCloseRequestResponse> {
    const response = await fetch(`${API_BASE_URL}/client/close-requests/${closeRequestId}/reject`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Failed to reject close request' }));
        throw new Error(error.message || 'Failed to reject close request');
    }

    return response.json();
}

